File: services/inference_worker/inference_worker/stages/line_crossing.py
"""Stage 5 -- Line crossing detection.

Tests whether a track's centroid has crossed any configured counting line
between the current and previous frame using segment intersection (cross-product
method).  Handles deduplication via per-track crossing sequence numbers and
cooldown windows.
"""

from __future__ import annotations

import logging
from typing import TYPE_CHECKING

from inference_worker.models import CrossingResult, TrackState
from shared_contracts.geometry import CountingLine, Point2D

if TYPE_CHECKING:
    from inference_worker.settings import CrossingSettings

logger = logging.getLogger(__name__)


def _cross_product(o: tuple[float, float], a: tuple[float, float], b: tuple[float, float]) -> float:
    """2D cross product of vectors (OA) x (OB)."""
    return (a[0] - o[0]) * (b[1] - o[1]) - (a[1] - o[1]) * (b[0] - o[0])


def _segments_intersect(
    p1: tuple[float, float],
    p2: tuple[float, float],
    q1: tuple[float, float],
    q2: tuple[float, float],
) -> bool:
    """Test whether line segments (p1-p2) and (q1-q2) intersect."""
    d1 = _cross_product(q1, q2, p1)
    d2 = _cross_product(q1, q2, p2)
    d3 = _cross_product(p1, p2, q1)
    d4 = _cross_product(p1, p2, q2)

    if ((d1 > 0 and d2 < 0) or (d1 < 0 and d2 > 0)) and (
        (d3 > 0 and d4 < 0) or (d3 < 0 and d4 > 0)
    ):
        return True

    return False


def _determine_direction(
    prev: Point2D,
    curr: Point2D,
    line: CountingLine,
) -> str:
    """Determine crossing direction relative to the counting line's direction vector."""
    dx = curr.x - prev.x
    dy = curr.y - prev.y
    dot = dx * line.direction_vector.dx + dy * line.direction_vector.dy
    return "inbound" if dot > 0 else "outbound"


def check_single_crossing(
    prev_centroid: Point2D,
    curr_centroid: Point2D,
    counting_line: CountingLine,
) -> CrossingResult | None:
    """Check if the centroid path crosses a single counting line.

    Returns a CrossingResult if a valid crossing occurred, None otherwise.
    """
    p1 = (prev_centroid.x, prev_centroid.y)
    p2 = (curr_centroid.x, curr_centroid.y)
    q1 = (counting_line.start.x, counting_line.start.y)
    q2 = (counting_line.end.x, counting_line.end.y)

    if not _segments_intersect(p1, p2, q1, q2):
        return None

    direction = _determine_direction(prev_centroid, curr_centroid, counting_line)

    line_dir = counting_line.direction.lower()
    if line_dir != "bidirectional" and line_dir != direction:
        return None

    return CrossingResult(
        line_id=counting_line.name,
        line_name=counting_line.name,
        direction=direction,
    )


class LineCrossingDetector:
    """Stage 5: detect counting-line crossings for all active tracks."""

    def __init__(self, settings: CrossingSettings) -> None:
        self._settings = settings

    def check_crossings(
        self,
        track: TrackState,
        counting_lines: list[CountingLine],
        frame_index: int,
    ) -> list[CrossingResult]:
        """Check a single track against all counting lines.

        Applies cooldown deduplication and minimum displacement filtering.
        Returns a list of confirmed crossings (usually 0 or 1).
        """
        if len(track.centroid_history) < 2:
            return []

        prev = track.centroid_history[-2]
        curr = track.centroid_history[-1]

        displacement = ((curr.x - prev.x) ** 2 + (curr.y - prev.y) ** 2) ** 0.5
        if displacement < self._settings.min_displacement:
            return []

        results: list[CrossingResult] = []

        for line in counting_lines:
            line_id = line.name

            last_frame = track.last_crossing_frame.get(line_id, -999)
            if (frame_index - last_frame) < self._settings.cooldown_frames:
                continue

            crossing = check_single_crossing(prev, curr, line)
            if crossing is not None:
                seq = track.crossing_sequences.get(line_id, 0) + 1
                track.crossing_sequences[line_id] = seq
                track.last_crossing_frame[line_id] = frame_index
                results.append(crossing)

        return results

File: apps/mobile_flutter/lib/core/inference/line_crossing.dart
/// Stage 5 — Line crossing detection.
///
/// Tests whether a track's centroid has crossed any configured counting line
/// between the current and previous frame using segment intersection
/// (cross-product method). Handles deduplication via per-track crossing
/// sequence numbers and cooldown windows.
///
/// Ported from `services/inference_worker/inference_worker/stages/line_crossing.py`.
library;

import 'dart:math' as math;

import 'package:greyeye_mobile/core/inference/models.dart';
import 'package:greyeye_mobile/core/inference/pipeline_settings.dart';

class LineCrossingDetector {
  LineCrossingDetector(this._settings);

  final CrossingSettings _settings;

  /// Check a single track against all counting lines.
  ///
  /// Applies cooldown deduplication and minimum displacement filtering.
  /// Returns a list of confirmed crossings (usually 0 or 1).
  List<CrossingResult> checkCrossings(
    TrackState track,
    List<CountingLine> countingLines,
    int frameIndex,
  ) {
    if (track.centroidHistory.length < 2) return [];

    final prev = track.centroidHistory[track.centroidHistory.length - 2];
    final curr = track.centroidHistory.last;

    final displacement = math.sqrt(
      math.pow(curr.x - prev.x, 2) + math.pow(curr.y - prev.y, 2),
    );
    if (displacement < _settings.minDisplacement) return [];

    final results = <CrossingResult>[];

    for (final line in countingLines) {
      final lineId = line.name;

      final lastFrame = track.lastCrossingFrame[lineId] ?? -999;
      if ((frameIndex - lastFrame) < _settings.cooldownFrames) continue;

      final crossing = _checkSingleCrossing(prev, curr, line);
      if (crossing != null) {
        final seq = (track.crossingSequences[lineId] ?? 0) + 1;
        track.crossingSequences[lineId] = seq;
        track.lastCrossingFrame[lineId] = frameIndex;
        results.add(crossing);
      }
    }

    return results;
  }
}

CrossingResult? _checkSingleCrossing(
  Point2D prevCentroid,
  Point2D currCentroid,
  CountingLine countingLine,
) {
  final p1 = (prevCentroid.x, prevCentroid.y);
  final p2 = (currCentroid.x, currCentroid.y);
  final q1 = (countingLine.start.x, countingLine.start.y);
  final q2 = (countingLine.end.x, countingLine.end.y);

  if (!_segmentsIntersect(p1, p2, q1, q2)) return null;

  final direction =
      _determineDirection(prevCentroid, currCentroid, countingLine);

  final lineDir = countingLine.direction.toLowerCase();
  if (lineDir != 'bidirectional' && lineDir != direction) return null;

  return CrossingResult(
    lineId: countingLine.name,
    lineName: countingLine.name,
    direction: direction,
  );
}

/// 2D cross product of vectors (OA) × (OB).
double _crossProduct(
  (double, double) o,
  (double, double) a,
  (double, double) b,
) {
  return (a.$1 - o.$1) * (b.$2 - o.$2) - (a.$2 - o.$2) * (b.$1 - o.$1);
}

/// Test whether line segments (p1–p2) and (q1–q2) properly intersect.
bool _segmentsIntersect(
  (double, double) p1,
  (double, double) p2,
  (double, double) q1,
  (double, double) q2,
) {
  final d1 = _crossProduct(q1, q2, p1);
  final d2 = _crossProduct(q1, q2, p2);
  final d3 = _crossProduct(p1, p2, q1);
  final d4 = _crossProduct(p1, p2, q2);

  if (((d1 > 0 && d2 < 0) || (d1 < 0 && d2 > 0)) &&
      ((d3 > 0 && d4 < 0) || (d3 < 0 && d4 > 0))) {
    return true;
  }

  return false;
}

/// Determine crossing direction relative to the counting line's direction vector.
String _determineDirection(
  Point2D prev,
  Point2D curr,
  CountingLine line,
) {
  final dx = curr.x - prev.x;
  final dy = curr.y - prev.y;

  // Use explicit direction vector if available, otherwise derive a perpendicular
  // from the line endpoints (rotated 90° clockwise).
  final dirVec = line.directionVector;
  final dvx = dirVec?.x ?? (line.end.y - line.start.y);
  final dvy = dirVec?.y ?? -(line.end.x - line.start.x);

  final dot = dx * dvx + dy * dvy;
  return dot > 0 ? 'inbound' : 'outbound';
}
